bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/link/ItemChannelLinkRegistry.java

```diff
@@ -99,19 +198,40 @@ public void removeLinksForThing(ThingUID thingUID) {
 
     @Override
     protected void notifyListenersAboutAddedElement(ItemChannelLink element) {
-        super.notifyListenersAboutAddedElement(element);
-        postEvent(LinkEventFactory.createItemChannelLinkAddedEvent(element));
+        notifyListenersAboutAddedElement(element, itemRegistry.containsItem(element.getItemName()));
     }
 
     @Override
     protected void notifyListenersAboutRemovedElement(ItemChannelLink element) {
-        super.notifyListenersAboutRemovedElement(element);
-        postEvent(LinkEventFactory.createItemChannelLinkRemovedEvent(element));
+        // Workaround for removeProvider: see description of member 'active' above
+        final ItemRegistry itemRegistry = this.itemRegistry;
+        final boolean contains;
+        if (itemRegistry != null) {
+            contains = itemRegistry.containsItem(element.getItemName());
+        } else {
+            contains = true;
+        }
+        notifyListenersAboutRemovedElement(element, contains);
     }
 
     @Override
     protected void notifyListenersAboutUpdatedElement(ItemChannelLink oldElement, ItemChannelLink element) {
         super.notifyListenersAboutUpdatedElement(oldElement, element);
         // it is not needed to send an event, because links can not be updated
     }
+
+    protected void notifyListenersAboutAddedElement(final ItemChannelLink element, boolean itemExists) {
+        if (itemExists) {
+            super.notifyListenersAboutAddedElement(element);
```

Comment:
> Should we inform the listeners on link creation and item creation

I would say only on link creation, because this is what they register for -> keep the behavior on this part as it was.

> Perhaps I misunderstand your comments I assumed you want to use that events in the channel link to refresh the channels, too.

Right, which you correctly implemented in https://github.com/eclipse/smarthome/pull/2122/commits/cbe1eb15fde14ab92eef250b92d6ced8c15b4ec5 - so I would suggest you keep this code in place as well.

> Can we send the event twice on link creation and on item creation?

No, but the ThinkLinkManager can register on both evens (links and items) and call the "channelLinked()" method when it is sure that both link+item exist.